src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,473 @@
+      @Nonnull final Set<String> productTypeIds,
+      @Nonnull final Set<String> customObjectIds) {
+
+    final Set<String> nonCachedProductIds = getNonCachedReferenceIds(productIds);
+    final Set<String> nonCachedCategoryIds = getNonCachedReferenceIds(categoryIds);
+    final Set<String> nonCachedProductTypeIds = getNonCachedReferenceIds(productTypeIds);
+    final Set<String> nonCachedCustomObjectIds = getNonCachedReferenceIds(customObjectIds);
+
+    if (nonCachedProductIds.isEmpty()
+        && nonCachedCategoryIds.isEmpty()
+        && nonCachedProductTypeIds.isEmpty()) {
+      return fetchCustomObjectKeys(customObjectIds);
+    }
+
+    List<List<String>> productIdsChunk = ChunkUtils.chunk(nonCachedProductIds, CHUNK_SIZE);
+    List<List<String>> categoryIdsChunk = ChunkUtils.chunk(nonCachedCategoryIds, CHUNK_SIZE);
+    List<List<String>> productTypeIdsChunk = ChunkUtils.chunk(nonCachedProductTypeIds, CHUNK_SIZE);
+
+    List<ResourceIdsGraphQlRequest> collectedRequests = new ArrayList<>();
+
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(productIdsChunk, GraphQlQueryResources.PRODUCTS));
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(categoryIdsChunk, GraphQlQueryResources.CATEGORIES));
+    collectedRequests.addAll(
+        createResourceIdsGraphQlRequests(productTypeIdsChunk, GraphQlQueryResources.PRODUCT_TYPES));
+
+    return ChunkUtils.executeChunks(getCtpClient(), collectedRequests)
+        .thenApply(ChunkUtils::flattenGraphQLBaseResults)
+        .thenApply(
+            results -> {
+              cacheResourceReferenceKeys(results);
+              return referenceIdToKeyCache;
+            })
+        .thenCompose(ignored -> fetchCustomObjectKeys(nonCachedCustomObjectIds));
+  }
+
+  @Nonnull
+  private CompletableFuture<Map<String, String>> fetchCustomObjectKeys(
+      @Nonnull final Set<String> nonCachedCustomObjectIds) {
+
+    if (nonCachedCustomObjectIds.isEmpty()) {
+      return CompletableFuture.completedFuture(referenceIdToKeyCache);
+    }
+
+    final List<List<String>> chunkedIds = ChunkUtils.chunk(nonCachedCustomObjectIds, CHUNK_SIZE);
+
+    final List<CustomObjectQuery<JsonNode>> chunkedRequests =
+        chunkedIds.stream()
+            .map(
+                ids ->
+                    CustomObjectQuery.ofJsonNode()
+                        .plusPredicates(p -> p.id().isIn(ids))
+                        .withLimit(CHUNK_SIZE)
+                        .withFetchTotal(false))
+            .collect(toList());
+
+    return ChunkUtils.executeChunks(getCtpClient(), chunkedRequests)
+        .thenApply(ChunkUtils::flattenPagedQueryResults)
+        .thenApply(
+            customObjects -> {
+              customObjects.forEach(
+                  customObject -> {
+                    referenceIdToKeyCache.put(
+                        customObject.getId(),
+                        CustomObjectCompositeIdentifier.of(customObject).toString());
```

Comment:
I checked if this would be needed a placeholder here, but in commercetools, both container and key are required fields, looks like we are safe in this case, so looks like no additional check is needed.